File: lib/md2pdf/footnotes.lua
-- Collects every Note inline in the document, replaces it with a
-- numbered <sup> reference, and appends a section of definitions
-- with backrefs at the end. The section heading text is taken
-- from the `footnotes-title` metadata (default `Footnotes`). When
-- the input markdown already has a manual heading immediately
-- before the footnote definitions (e.g. `## Footnotes`), pandoc
-- consumes the definitions and leaves the heading dangling at the
-- end of the document; that trailing heading is removed here so
-- it does not double up with the auto-rendered section.

local function meta_string(meta, key, default)
  if meta[key] == nil then return default end
  return pandoc.utils.stringify(meta[key])
end

local notes = {}
local notes_by_key = {}
local first_ref_anchor = {}
local ref_counter = 0

function Note(elem)
  local key = pandoc.utils.stringify(elem.content)
  local id = notes_by_key[key]
  if not id then
    id = #notes + 1
    notes[id] = elem.content
    notes_by_key[key] = id
  end

  ref_counter = ref_counter + 1
  local anchor = 'fnref' .. id .. '-' .. ref_counter
  if not first_ref_anchor[id] then
    first_ref_anchor[id] = anchor
  end

  local link = pandoc.Link(
    { pandoc.Str(tostring(id)) },
    '#fn' .. id,
    '',
    pandoc.Attr(
      anchor,
      { 'footnote-ref' },
      { { 'role', 'doc-noteref' } }
    )
  )
  return pandoc.Superscript({ link })
end

local function strip_trailing_heading(blocks)
  for i = #blocks, 1, -1 do
    local b = blocks[i]
    if b.t == 'Header' then
      table.remove(blocks, i)
      return
    end
    if b.t ~= 'Null' then return end
  end
end

local function append_backref(item_blocks, anchor)
  for i = #item_blocks, 1, -1 do
    local b = item_blocks[i]
    if b.t == 'Para' or b.t == 'Plain' then
      table.insert(b.content, pandoc.Space())
      table.insert(b.content, pandoc.Link(
        { pandoc.Str('\u{21A9}') },
        '#' .. anchor,
        '',
        pandoc.Attr(
          '',
          { 'footnote-back' },
          { { 'role', 'doc-backlink' } }
        )
      ))
      return
    end
  end
  table.insert(item_blocks, pandoc.Para({
    pandoc.Link(
      { pandoc.Str('\u{21A9}') },
      '#' .. anchor,
      '',
      pandoc.Attr(
        '',
        { 'footnote-back' },
        { { 'role', 'doc-backlink' } }
      )
    )
  }))
end

local function render_item(blocks)
  local doc = pandoc.Pandoc(blocks)
  local html = pandoc.write(doc, 'html5')
  return html
end

function Pandoc(doc)
  if #notes == 0 then return doc end

  local title = meta_string(doc.meta, 'footnotes-title', 'Footnotes')

  strip_trailing_heading(doc.blocks)

  if title ~= '' then
    table.insert(doc.blocks, pandoc.Header(
      2,
      { pandoc.Str(title) },
      pandoc.Attr('footnotes', { 'footnotes-title' }, {})
    ))
  end

  local html_parts = {
    '<div id="footnotes-section" class="footnotes" role="doc-endnotes">',
    '<ol class="footnotes-list">'
  }
  for i, blocks in ipairs(notes) do
    local item = {}
    for _, b in ipairs(blocks) do table.insert(item, b) end
    append_backref(item, first_ref_anchor[i])
    html_parts[#html_parts + 1] =
      '<li id="fn' .. i .. '" role="doc-endnote">' ..
      render_item(item) .. '</li>'
  end
  html_parts[#html_parts + 1] = '</ol></div>'

  table.insert(
    doc.blocks,
    pandoc.RawBlock('html', table.concat(html_parts, '\n'))
  )
  return doc
end

File: lib/transpareo/md2pdf/filters/footnotes.rb
# frozen_string_literal: true

require 'cgi'

module Transpareo
  module Md2pdf
    module Filters
      # Rebuilds the footnote apparatus.
      #
      # commonmarker already parses `[^id]` references and gathers
      # the definitions, but it numbers them per label. This filter
      # renumbers in reference order and merges entries whose
      # definition text is identical, so citing the same source from
      # three places yields one numbered entry rather than three.
      #
      # A manual `## Footnotes` heading at the end of the source is
      # consumed here, so it does not double up with the heading
      # this filter renders.
      module Footnotes
        # U+21A9, the conventional footnote backlink glyph.
        BACKREF = "\u21A9"

        module_function

        def call(doc)
          section = doc.fragment.at_css('section[data-footnotes]')
          return unless section

          definitions = extract_definitions(section)
          section.remove
          return if definitions.empty?

          items = renumber(doc, definitions)
          return if items.empty?

          strip_trailing_heading(doc)
          append_section(doc, items)
        end

        # Maps each definition's element id to its rendered body and
        # a normalised key used to spot duplicates.
        def extract_definitions(section)
          section.css('li').each_with_object({}) do |li, out|
            id = li['id'].to_s
            next if id.empty?

            li.css('a.footnote-backref').each(&:remove)
            html = li.inner_html.strip
            out[id] = { html: html, key: normalize(li.text) }
          end
        end

        def normalize(text)
          text.gsub(/\s+/, ' ').strip
        end

        # Walks the references in document order, assigning numbers
        # as new definition texts are first seen.
        def renumber(doc, definitions)
          items = []
          by_key = {}
          counter = 0

          doc.fragment.css('sup.footnote-ref').each do |sup|
            link = sup.at_css('a')
            next unless link

            definition = definitions[link['href'].to_s.delete_prefix('#')]
            next unless definition

            index = by_key[definition[:key]]
            unless index
              items << { html: definition[:html], anchor: nil }
              index = items.size
              by_key[definition[:key]] = index
            end

            counter += 1
            anchor = "fnref#{index}-#{counter}"
            items[index - 1][:anchor] ||= anchor
            sup.replace(reference_markup(index, anchor))
          end

          items
        end

        def reference_markup(index, anchor)
          %(<sup class="footnote-ref"><a id="#{anchor}" ) +
            %(href="#fn#{index}" role="doc-noteref">#{index}</a></sup>)
        end

        # Removes a heading left dangling at the end of the document
        # once the definitions that followed it were gathered into
        # the footnotes section.
        def strip_trailing_heading(doc)
          last = doc.fragment.children.reverse.find do |node|
            node.element? || (node.text? && !node.text.strip.empty?)
          end
          last.remove if last&.element? && last.name.match?(/\Ah[1-6]\z/)
        end

        def append_section(doc, items)
          label = doc[:footnotes_label]
          unless label.to_s.empty?
            text = CGI.escapeHTML(label.to_s)
            doc.fragment.add_child(
              %(<h2 id="footnotes" class="footnotes-title">#{text}</h2>)
            )
          end
          doc.fragment.add_child(section_markup(items))
        end

        def section_markup(items)
          parts = [
            '<div id="footnotes-section" class="footnotes" ' \
            'role="doc-endnotes">',
            '<ol class="footnotes-list">'
          ]
          items.each_with_index do |item, i|
            parts << list_item(item, i + 1)
          end
          parts << '</ol></div>'
          parts.join("\n")
        end

        def list_item(item, index)
          back = %(<a href="##{item[:anchor]}" class="footnote-back" ) +
                 %(role="doc-backlink">#{BACKREF}</a>)
          %(<li id="fn#{index}" role="doc-endnote">) +
            "#{merge_backref(item[:html], back)}</li>"
        end

        # The backref belongs inside the last paragraph so it sits on
        # the same line as the closing text rather than below it.
        # Removing the original backref leaves whitespace behind, so
        # the tail is trimmed before the new one is appended.
        def merge_backref(html, back)
          if html.end_with?('</p>')
            "#{html[0...-4].rstrip} #{back}</p>"
          else
            "#{html.rstrip} #{back}"
          end
        end
      end
    end
  end
end
